### PR TITLE
feat(coro): coro 2.0

### PR DESCRIPTION
**Pull request by Mishura4**
This is a *massive* PR that fundamentally changes the coro implementation, everything is a derivative of dpp::awaitable and dpp::promise now. This helps with de-duplicating a lot of coro-related code, making it easier to maintain, and offers a way to add things to all of the coroutine types at once, for example the `sync_wait` method available on all awaitable types, which can help for example with starting a coroutine from a non-coroutine.

Still missing some documentation & polishing

## Code change checklist

- [ ] I have ensured that all methods and functions are **fully documented** using doxygen style comments.
- [x] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [x] I tested that my change works before raising the PR.
- [x] I have ensured that I did not break any existing API calls.
- [x] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

### CODE REVIEW

**Comment by Jaskowicz1** [comment]
This is huge, god damn.

I'll pull this branch locally and test this with my bot, will give you a shout on Discord if I have any issues!

**Comment by braindigitalis** [comment]
lgtm, will test it out when its in dev as the coro features in my bot arent complete so i can't test this thoroughly.

File: include/dpp/coro/awaitable.h [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+		});
+	}
+
+	/**
+	 * @brief Blocks this thread and waits for the awaitable to finish.
+	 *
+	 * @attention This will BLOCK THE THREAD. It is likely you want to use co_await instead.
+	 * @param duration Maximum duration to wait for
+	 * @retval If T is void, returns a boolean for which true means the awaitable completed, false means it timed out.
+	 * @retval If T is non-void, returns a std::optional<T> for which an absense of value means timed out.
```
Comment (Mishura4) [inline include/dpp/coro/awaitable.h:149]: note to self: fix "absence"

File: include/dpp/coro/task.h [hunk 2]
```diff
@@ -375,136 +134,56 @@ class task : private detail::task::task_base<R> {
...
 	 *
-	 * @throw On resumption, any exception thrown by the coroutine is propagated to the caller.
-	 * @return On resumption, this expression evaluates to the result object of type R, as a const reference.
+	 * Destroys the handle. If the task is still running, it will be cancelled.
 	 */
-	[[nodiscard]] const auto& operator co_await() const & noexcept {
-		return static_cast<const detail::task::task_base<R>&>(*this);
+	~task() {
+		if (handle && this->valid()) {
+			if (this->abandon() & state_flags::sf_done) {
```
Comment (Mishura4) [inline include/dpp/coro/task.h:150]: note to self: don't use an implicit conversion to bool

Comment (Jaskowicz1) [reply include/dpp/coro/task.h:150]: note to self: note

**Comment by Mishura4** [comment]
Closing to reopen from https://github.com/brainboxdotcc/DPP/tree/coro

**Pull request closed without merging**